### DISCUSSION TITLE
Confusing results with `brew deps`

### DISCUSSION THREAD

**Discussion by dr-scsi** [Homebrew/discussions#5148] [Everyday usage] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.2.9
ORIGIN: https://github.com/Homebrew/brew
HEAD: e5fefd73cd97cd36ae3af29551f529ae59b333d6
Last commit: 3 days ago
Core tap JSON: 22 Feb 08:01 UTC
Core cask tap JSON: 22 Feb 08:01 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 3.1.4 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: octa-core 64-bit arm_blizzard_avalanche
Clang: 15.0.0 build 1500
Git: 2.43.2 => /opt/homebrew/bin/git
Curl: 8.4.0 => /usr/bin/curl
macOS: 14.3.1-arm64
CLT: 15.1.0.0.1.1700200546
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Putting non-prefixed coreutils in your path can cause GMP builds to fail.
```

### Description of issue

I see different results when using `brew deps`, e.g.:

```
$ brew deps ldns
ca-certificates
mpdecimal
openssl@3
python@3.11
readline
sqlite
xz
```
vs.

```
-> brew deps --tree ldns
ldns
├── openssl@3
│   └── ca-certificates
└── python@3.12
    ├── mpdecimal
    ├── openssl@3
    │   └── ca-certificates
    ├── sqlite
    │   └── readline
    └── xz
```

The required python version is 3.11 or 3.12 depending on the `--tree` key.  Am I missing something?

**Comment by SMillerDev** [comment]
Could it be that you ran `brew update` in between? We switched the default python recently.

**Comment by dr-scsi** [comment]
@SMillerDev - Thanks for your response.  No, I didn't run `brew update` in between:
```
$ brew update
Already up-to-date.

$ brew deps ldns
ca-certificates
mpdecimal
openssl@3
python@3.11
readline
sqlite
xz

$ brew deps --tree ldns
ldns
├── openssl@3
│   └── ca-certificates
└── python@3.12
    ├── mpdecimal
    ├── openssl@3
    │   └── ca-certificates
    ├── sqlite
    │   └── readline
    └── xz
```

Reg. default python: Yes, I saw that, and I wanted to uninstall python@3.11, but I get:
```
$ brew uninstall python@3.11
Error: Refusing to uninstall /opt/homebrew/Cellar/python@3.11/3.11.7_2
because it is required by ldns, which is currently installed.
You can override this and force removal with:
  brew uninstall --ignore-dependencies python@3.11
```
This is where the exercise above began.

**Comment by gromgit** [comment, marked as answer]
I seem to recall one of the Homebrew maintainers mentioning quite some time back (can't find the exact post now) that `brew deps <formula>` and `brew deps --tree <formula>` actually take their info from different sources _if `<formula>` is installed_. If my memory serves me correctly:
* `brew deps <formula>` reads `INSTALL_RECEIPT.json` to see what the dependencies were _at the time of installation_
  * this would, among other things, indicate which formulae could not be uninstalled without breaking `<formula>`
* `brew deps --tree <formula>` reads the current state of Homebrew core to see what the dependencies are _if you (re)install or upgrade `<formula>`_
  * this would, among other things, indicate which other formulae need to be upgraded or installed fresh

**Reply by ZhongRuoyu** [reply]
> `brew deps <formula>` and `brew deps --tree <formula>` actually take their info from different sources _if `<formula>` is installed_

Here they are:

- https://github.com/Homebrew/brew/issues/16032#issuecomment-1734885954
- https://github.com/orgs/Homebrew/discussions/4784#discussioncomment-7028546
- https://github.com/Homebrew/brew/issues/13717#issuecomment-1246717893

**Comment by dr-scsi** [comment]
@gromgit - Thanks, I think your memory is serving you correctly 👍  I did:

```
$ brew reinstall ldns

$ brew deps ldns
ca-certificates
mpdecimal
openssl@3
python@3.12
readline
sqlite
xz
```
Now the python version is updated.  I wonder if this should be documented somewhere.  But the issue resolved, thanks.
